pkg/inclusion/paths.go

```diff
@@ -0,0 +1,113 @@
+package inclusion
+
+// genSubTreeRootPath calculates the path to a given subtree root of a node, given the
+// depth and position of the node. note: the root of the tree is depth 0.
+// nolint
+func genSubTreeRootPath(depth int, pos uint) []bool {
+	path := make([]bool, depth)
+	for i := depth; i >= 0; i-- {
+		if (pos & (1 << i)) == 0 {
+			path = append(path, false)
+		} else {
+			path = append(path, true)
+		}
+	}
+	return path
+}
+
+// coord identifies a tree node using the depth and position
+type coord struct {
+	// depth is the typical depth of a tree, 0 being the root
+	depth uint64
+	// position is the index of a node of a given depth, 0 being the left most
+	// node
+	position uint64
+}
+
+// climb is a state transition function to simulate climbing a balanced binary
+// tree, using the current node as input and the next highest node as output.
+func (c coord) climb() coord {
+	return coord{
+		depth:    c.depth - 1,
+		position: c.position / 2,
+	}
+}
+
+// canClimbRight uses the current position to calculate the direction of the next
+// climb. Returns true if the next climb is right (if the position (index) is
+// even).
+func (c coord) canClimbRight() bool {
+	return c.position%2 == 0
+}
```

Comment:
[no change needed] one alternative way of describe this relationship is parent / child 

```golang
// getParent returns the coordinate for the parent of the current coordinate.
func (c coord) getParent() coord {
	return coord{
		depth:    c.depth - 1,
		position: c.position / 2,
	}
}

// isLeftChild returns ture if the coordinate are the left child of a parent.
func (c coord) isLeftChild() bool {
	return c.position%2 == 0
}
```

but I think the climb terminology makes more sense given it's usage below.

Reply:
>but I think the climb terminology makes more sense given it's usage below

yeah, this is precisely the reason why I initially went with climb.